src/commands.rs

```diff
@@ -0,0 +1,455 @@
+            children: Vec::new(),
+            data: NodeData::Argument(Argument {
+                name: Default::default(),
+                parser: Parser::BrigadierInteger(BrigadierInteger {
+                    min: None,
+                    max: None,
+                }),
+                suggestions_type: None
+            }),
+            is_executable: false,
+            redirect_node: None,
+        });
+        /*match command_registry.cache.entry(TypeId::of::<Self>()) {
+            Entry::Vacant(vac) => {
+                vac.insert(id);
+                command_registry.graph.push(Node {
+                    children: Vec::new(),
+                    data: NodeData::Argument(Argument {
+                        name: Default::default(),
+                        parser: Parser::BrigadierInteger(BrigadierInteger {
+                            min: None,
+                            max: None,
+                        }),
+                        suggestions_type: None
+                    }),
+                    is_executable: false,
+                    redirect_node: None,
+                });
+            }
+            Entry::Occupied(occ) => {
+                command_registry.graph.push(Node {
+                    children: Vec::new(),
+                    data: NodeData::Argument(Argument {
+                        name: Default::default(),
+                        parser: Parser::BrigadierInteger(BrigadierInteger {
+                            min: None,
+                            max: None,
+                        }),
+                        suggestions_type: None
+                    }),
+                    is_executable: false,
+                    redirect_node: Some(occ.get().clone()),
+                });
+            }
+        };*/
+        id
+    }
+}
+
+#[derive(Debug)]
+pub struct CommandRegistry {
+    pub cache: BTreeMap<TypeId, VarInt>,
+    pub graph: Vec<Node>,
+}
+
+impl CommandRegistry {
+    pub fn new() -> Self {
+        CommandRegistry {
+            cache: Default::default(),
+            graph: vec![
+                Node {
+                    children: vec![],
+                    data: NodeData::Root,
+                    is_executable: false,
+                    redirect_node: None
+                }
+            ],
+        }
+    }
+
+    pub fn register_command<C: AddToCommands>(&mut self) {
+        let id = C::add_to_commands(self);
+        self.graph[0].children.push(id);
+    }
+
+    pub fn to_packet(&self) -> s2c::play::Commands {
+        s2c::play::Commands {
+            nodes: self.graph.clone(),
+            root_index: VarInt(0),
+        }
+    }
+}
+
+pub trait AddToCommands {
+    fn add_to_commands(commands: &mut CommandRegistry) -> VarInt;
+}
+
+macro_rules! if_typ_is_empty_expr {
+    (, $t:expr, $f:expr) => {
+        $t
+    };
+    ($typ:ty, $t:expr, $f:expr) => {
+        $f
+    };
+}
+
+#[macro_export]
+macro_rules! def_command {
```

Comment:
We should avoid exposing a macro based solution. Remember that Valence is supposed to be minimal and should only contain what is absolutely necessary to work with the protocol. Extra convenience stuff should be in another crate.

Besides, what I think users would want instead is a [derive macro](https://doc.rust-lang.org/reference/procedural-macros.html#derive-macros), and proc macros need to be in a separate crate anyway.

Reply:
I should have discussed this beforehand, true.. I though about making a proc macro, but thought maybe it would work as is. But you're absolutely right. Another re-write will happen then eventually :P